### PR TITLE
Fix overrideStatements for insert actions. Fixes #1850

### LINKED ISSUE #1850: overrideStatements doesn't work with forceInsertAll
I'm trying to write a bulk upsert function, and my test for it is failing due to primary key conflicts

```scala
  def upsertWithPkey[E, T <: Table[E]](
      keyColumns: Iterable[String],
      batch: Seq[E]
  )(
      table: TableQuery[T]
  )(
      implicit ec: ExecutionContext
  ): DBIO[UpsertResult] = {
    val insertAll = table.forceInsertAll(batch)
    val insertStatement: String = insertAll.statements.headOption.getOrElse(
      throw new IllegalStateException(
        "Insert action has more than one statement."))

    val pkeyFragment = keyColumns.mkString(", ")
    val columns = table.baseTableRow.create_*.map(_.name)
    val setFragment = (columns.toSet -- keyColumns.toSet)
      .map(c => s"$c = EXCLUDED.$c")
      .mkString(", ")
    val upsertFragment =
      s"""
         | ON CONFLICT ($pkeyFragment) DO UPDATE
         |SET $setFragment
       """.stripMargin

    val operation =
      insertAll.overrideStatements(Seq(insertStatement + upsertFragment))
    operation.map {
      case None =>
        // http://books.underscore.io/essential-slick/essential-slick-3.html#inserting-multiple-rows
        //   [...] the type is Option[Int]. The reason for this is that the JDBC specification permits
        //   the underlying database driver to indicate that the number of rows inserted is unknown.
        UpsertResult(0, batch.size) // Make it up rather than crashing.
      case Some(insertedCount) =>
        UpsertResult(rowsUpdated = batch.size - insertedCount,
                     rowsInserted = insertedCount)
    }
  }
```

Specifically
```scala
    val operation =
      insertAll.overrideStatements(Seq(insertStatement + upsertFragment))
```

---

When I run it in my tests with logging on, I get 
```
[DEBUG] [action] #2: : MultiInsertAction 
    : [insert into "test_table" ("id","a","b")  values (?,?,?)
    :  ON CONFLICT (id) DO UPDATE
    : SET a = EXCLUDED.a, b = EXCLUDED.b
    :        ]
[DEBUG] [statement] Preparing statement: insert into "test_table" ("id","a","b")  values (?,?,?)
```

### PR DESCRIPTION
**Pull request by MichalPawlicki**
Fixes `overrideStatements` method for `MultiInsertAction` and `SingleInsertAction`. Fixes https://github.com/slick/slick/issues/1850.

### CODE REVIEW

**Comment by SethTisue** [comment]
if this is still relevant, rebase on current master? then we can try again to find a reviewer

**Comment by MichalPawlicki** [comment]
@SethTisue I think this is still relevant. Rebased on master.

**Comment by nafg** [comment]
Closing and reopening to trigger CI

File: slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala [hunk 1]
```diff
@@ -514,7 +514,7 @@ trait JdbcActionComponent extends SqlActionComponent { self: JdbcProfile =>
       session.withPreparedStatement(sql)(f)
 
     class SingleInsertAction(a: compiled.Artifacts, value: U) extends SimpleJdbcProfileAction[SingleInsertResult]("SingleInsertAction", Vector(a.sql)) {
-      def run(ctx: Backend#Context, sql: Vector[String]) = preparedInsert(a.sql, ctx.session) { st =>
+      def run(ctx: Backend#Context, sql: Vector[String]) = preparedInsert(sql.head, ctx.session) { st =>
```
Comment (nafg) [inline slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala:517]: Hi, I don't know the codebase well enough to understand this change. Can you explain what's happening? What is `a` and what gets passed as `sql` to `run`? Can you ELI5 why the old way is fails and the new way fixes it? Thanks.

File: slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMiscTest.scala [hunk 2]
```diff
@@ -57,16 +57,26 @@ class JdbcMiscTest extends AsyncTest[JdbcTestDB] {
...
     seq(
-      t.schema.create,
+      (t.schema ++ u.schema).create,
       t ++= Seq(1, 2, 3),
       a1.result.map(_ shouldBe Seq(1)),
       a1.result.overrideStatements(a2.result.statements).map(_ shouldBe Seq(2)),
       a1.result.head.map(_ shouldBe 1),
-      a1.result.head.overrideStatements(a2.result.head.statements).map(_ shouldBe 2)
+      a1.result.head.overrideStatements(a2.result.head.statements).map(_ shouldBe 2),
+      (t += 4).overrideStatements(Seq(u.insertStatement)),
```
Comment (nafg) [inline slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMiscTest.scala:76]: Can you add some comments to the code explaining what the test is doing? Thanks.

File: slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala (hunk 1 above)
Comment (MichalPawlicki) [reply slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala:517]: Hi, thanks for looking into this. I don't know the codebase well, either, but I'll try to explain. 🙂 

`SingleInsertAction` and `MultiInsertAction` extend `SimpleJdbcProfileAction`. The class has a complicated structure which seems to have been introduced to support statement overrides: https://github.com/slick/slick/blob/98ded3a777372d0783dbb42233b08ff51f6da3a7/slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala#L25-L32
It seems that `overrideStatements` depends on the fact that `run(ctx, sql)` actually uses `sql` to prepare a statement, and not the argument `a` passed to the constructor of `SingleInsertAction` or `MultiInsertAction`. 

Now, when the `overrideStatements` isn't called, `a.sql` ends up being passed to `run(ctx, sql)` anyway, because:
- `Vector(a.sql)` is passed to `SimpleJdbcProfileAction` as a constructor argument.
- `SimpleJdbcProfileAction#run(ctx)` calls `run(ctx, sql)` where `sql` is the value of this constructor argument.
- `run(ctx, sql)` is only ever called inside `SimpleJdbcProfileAction`. Its visibility could probably be changed to `protected` and it should still work.

Let me know if this answers your questions.

Comment (MichalPawlicki) [reply slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala:517]: If you think this would help, I can try to refactor these classes a bit.

File: slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMiscTest.scala (hunk 2 above)
Comment (nafg) [reply slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMiscTest.scala:78]: Ah, so the override is only for the "standard" part of the query (not counting the VALUES expression) and that is what `TableQuery[U].insertStatement` returns? Interesting.

I think I would find it less confusing if you just supplied a string instead of u.insertStatement, that way it's more focused on the behavior of overrideStatements and not on what insertStatement does, and I don't have to know what insertStatement does to understand the test.

Thoughts?

File: slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala (hunk 1 above)
Comment (nafg) [reply slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala:517]: Makes sense, thanks for the explanation!

File: slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMiscTest.scala [hunk 3]
```diff
@@ -74,13 +74,13 @@ class JdbcMiscTest extends AsyncTest[JdbcTestDB] {
       a1.result.head.map(_ shouldBe 1),
       a1.result.head.overrideStatements(a2.result.head.statements).map(_ shouldBe 2),
       /* Build an action that inserts a single value into table "t".
-         Then, override its statements with an "insert into u" statement. */
-      (t += 4).overrideStatements(Seq(u.insertStatement)),
+         Then, override its insert statement". */
```
Comment (nafg) [inline slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMiscTest.scala:77]:
```suggestion
         Then, override its insert statement. */
```

**Review by nafg — APPROVED** [review]

File: slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMiscTest.scala (hunk 2 above)
Comment (MichalPawlicki) [reply slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMiscTest.scala]: I think I'll have to revert to `u.insertStatement`. The tests are run on different DB engines and each of them supports a slightly different syntax (think ```insert into `u` ``` vs. `insert into "u"`).

Comment (MichalPawlicki) [reply slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMiscTest.scala:78]: Btw, `u.insertStatement` returns the whole statement (`insert into u (id) values (?)`). I updated the comment.

Comment (nafg) [reply slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMiscTest.scala:78]: What I'm confused about then is how the same statement works for inserting a single value and for inserting multiple rows. Not that has anything to do with your contribution, I'm just not clear how it works. Does ++= generate multiple insert statements? Or does the ? placeholder somehow support multiple rows?

**Review by nafg — APPROVED** [review]

File: slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMiscTest.scala (hunk 2 above)
Comment (MichalPawlicki) [reply slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMiscTest.scala:78]: AFAIK, Slick translates `++=` to JDBC code similar to this: https://stackoverflow.com/a/3786127. So it "prepares" a statement from a string with placeholders, and then adds values from each row as a separate batch.

Comment (nafg) [reply slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMiscTest.scala:78]: Interesting

**Comment by nafg** [comment]
Thanks!
